### ISSUE TITLE
return in finally can swallow exception

### ISSUE THREAD

**Issue by iritkatriel** [zopefoundation/Zope#1231]
In the [expr_match](https://github.com/zopefoundation/Zope/blob/072af95e25954837f2dc39ebeddf290042682144/src/OFS/FindSupport.py#L194) function:

```
def expr_match(ob, ed, c=InstanceDict, r=0):
    e, md, push, pop = ed
    push(c(ob, md))
    try:
        r = e.eval(md)
    finally:
        pop()
        return r
```

The `return` in the finally will swallow any in-flight exception. This means that if `r.eval(md)` raises any exception (including `KeyboardInterrupt`), this exception will be swallowed. This is probably not what the author intended. 

See also: https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions.

It is possible that the intention was to write:

```
def expr_match(ob, ed, c=InstanceDict, r=0):
    e, md, push, pop = ed
    push(c(ob, md))
    try:
        r = e.eval(md)
    finally:
        pop()
    return r
```

which would return r is there is no exception, but allow the exception to propagate if there is one.

**Comment by davisagli** [comment]
I agree this should probably be corrected, but in the case of an exception in the try block, I think this raises UnboundLocalError since `r` hasn't been assigned.

**Comment by iritkatriel** [comment]
> in the case of an exception in the try block, I think this raises UnboundLocalError since `r` hasn't been assigned.


In this case it wouldn't reach the 'return r', it will bubble the exception up.

**Comment by dataflake** [comment]
Solved by https://github.com/zopefoundation/Zope/commit/566ad553e95ce2abb1ae2bf07f50d2bf3c8c5f1b
